drivers/clock_control/nrf_power_clock.c

```diff
@@ -9,272 +9,296 @@
-	 * this function, allow it to run once. This is needed because if a
-	 * soft reset is triggered while watchdog is active, the LF clock will
-	 * already be running, but won't be configured yet (watchdog forces LF
-	 * clock to be running).
-	 *
-	 * That is, a hardware check won't work here, because even if the LF
-	 * clock is already running it might not be initialized. We need an
-	 * initialized flag.
-	 */
-
-	imask = irq_lock();
-
-	if (k32src_initialized) {
-		irq_unlock(imask);
-		goto lf_already_started;
-	}
-
-	k32src_initialized = 1U;
+	return 0;
+}
 
-	irq_unlock(imask);
+static const struct clock_control_driver_api clock_control_api = {
+	.on = clock_start,
+	.off = clock_stop,
+	.async_on = clock_async_start,
+	.get_status = get_status,
+};
 
-	/* Clear events if any */
-	NRF_CLOCK->EVENTS_LFCLKSTARTED = 0;
+static struct nrf_clock_control hfclk;
+static const struct nrf_clock_control_config hfclk_config = {
+	.start_tsk = NRF_CLOCK_TASK_HFCLKSTART,
+	.started_evt = NRF_CLOCK_EVENT_HFCLKSTARTED,
+	.stop_tsk = NRF_CLOCK_TASK_HFCLKSTOP
+};
 
-	/* Set LF Clock Source */
-	lf_clk_src = POINTER_TO_UINT(sub_system);
-	NRF_CLOCK->LFCLKSRC = lf_clk_src;
+DEVICE_AND_API_INIT(clock_nrf5_m16src,
+		    DT_INST_0_NORDIC_NRF_CLOCK_LABEL  "_16M",
+		    clock_control_init, &hfclk, &hfclk_config, PRE_KERNEL_1,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    &clock_control_api);
 
-#if defined(CONFIG_CLOCK_CONTROL_NRF_K32SRC_BLOCKING)
-	irq_disable(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
 
-	intenset = NRF_CLOCK->INTENSET;
-	nrf_clock_int_enable(NRF_CLOCK_INT_LF_STARTED_MASK);
+static struct nrf_clock_control lfclk;
+static const struct nrf_clock_control_config lfclk_config = {
+	.start_tsk = NRF_CLOCK_TASK_LFCLKSTART,
+	.started_evt = NRF_CLOCK_EVENT_LFCLKSTARTED,
+	.stop_tsk = NRF_CLOCK_TASK_LFCLKSTOP,
+	.stop_handler =
+		IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION) ?
+			nrf_clock_control_calibration_stop : NULL
+};
 
-	/* Start and spin-wait until clock settles */
-	nrf_clock_task_trigger(NRF_CLOCK_TASK_LFCLKSTART);
+DEVICE_AND_API_INIT(clock_nrf5_k32src,
+		    DT_INST_0_NORDIC_NRF_CLOCK_LABEL  "_32K",
+		    k32_dummy_init, &lfclk, &lfclk_config, PRE_KERNEL_1,
+		    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,
+		    &clock_control_api);
 
-	while (NRF_CLOCK->EVENTS_LFCLKSTARTED == 0) {
-		__WFE();
-		__SEV();
-		__WFE();
+static void clkstarted_handle(struct device *dev)
+{
+	struct clock_control_async_data *async_data;
+	struct nrf_clock_control *data = dev->driver_data;
+	sys_snode_t *node = sys_slist_get(&data->list);
+
+	DBG(dev, "Clock started");
+	data->started = true;
+
+	while (node != NULL) {
+		async_data = CONTAINER_OF(node,
+				struct clock_control_async_data, node);
+		async_data->cb(dev, async_data->user_data);
+		node = sys_slist_get(&data->list);
 	}
+}
 
-	NRF_CLOCK->EVENTS_LFCLKSTARTED = 0;
+#if defined(CONFIG_USB_NRF52840) || NRF_POWER_HAS_POFCON
+static bool power_event_check_and_clean(u32_t evt, u32_t intmask)
```

Comment:
By using `nrf_clock_event_t` for `evt` you could prevent such mistakes as the one with `nrf_clock_event_clear` below.

Reply:
ack